CMakeLists.txt

```diff
@@ -73,58 +44,99 @@ if("${isSystemDir}" STREQUAL "-1")
-# we will pick headers from a conda environment before our own. That's
-# very confusing!
-set(CMAKE_INCLUDE_DIRECTORIES_BEFORE ON)
+################################################################################
+# Third-party libraries
+################################################################################
 
-# Make it possible to start cmake directly in an appropriately conda environment
-if(NOT DEFINED CMAKE_PREFIX_PATH AND DEFINED ENV{CONDA_PREFIX})
-  set(CMAKE_PREFIX_PATH "$ENV{CONDA_PREFIX}")
+################################################################################
+# protobuf-3.5.2 (must use the same version as Caffe2 as long as we use Caffe2)
+################################################################################
+# TODO: turn this horror into a custom command and custom targets
+find_program(test_PROTOBUF_PROTOC_EXECUTABLE protoc
+  PATHS ${PROJECT_SOURCE_DIR}/third-party/googlelibraries/protobuf-3.5.2
+  PATH_SUFFIXES src/
+  NO_DEFAULT_PATH)
+if (NOT test_PROTOBUF_PROTOC_EXECUTABLE)
+  execute_process(
+    COMMAND ./autogen.sh
+    WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}/third-party/googlelibraries/protobuf-3.5.2
+  )
+  execute_process(
+    COMMAND ./configure "CFLAGS=-fPIC" "CXXFLAGS=-fPIC" --disable-shared --enable-static
+    WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}/third-party/googlelibraries/protobuf-3.5.2
+  )
 endif()
 
-# NB: If the finding logic here gets complicated, make a proper FindXXX.cmake
-# module and use find_package
-
-# Normal dependencies
-
-# pybind11
-include_directories(third-party/pybind11)
-
-# isl
-find_path(ISL_INCLUDE_DIR NAMES isl)
-include_directories(${ISL_INCLUDE_DIR})
-message(STATUS "PROJECT_SOURCE_DIR: ${PROJECT_SOURCE_DIR}")
-find_library(ISL_LIBRARIES isl PATHS ${CMAKE_INSTALL_PREFIX} PATH_SUFFIXES lib lib64)
-message(STATUS "Found isl: ${ISL_LIBRARIES}")
-
-# dlpack
-find_path(ISL_INCLUDE_DIR NAMES dlpack)
-include_directories(${DLPACK_INCLUDE_DIR})
-
-set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
-execute_process(COMMAND ${CLANG_PREFIX}/bin/llvm-config --libdir  OUTPUT_VARIABLE LLVM_LIBDIR OUTPUT_STRIP_TRAILING_WHITESPACE)
-link_directories(${LLVM_LIBDIR})
-
-# protobuf
-# TODO: check if this respects CMAKE_PREFIX_PATH
-find_package(Protobuf REQUIRED)
+# TODO: For some reason make -j"$(nproc)" fails here
+execute_process(
+  COMMAND make -j8
+  WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}/third-party/googlelibraries/protobuf-3.5.2
+)
+find_library(PROTOBUF_LIBRARIES libprotobuf.a
+  REQUIRED
+  PATHS ${PROJECT_SOURCE_DIR}/third-party/googlelibraries/protobuf-3.5.2
+  PATH_SUFFIXES src/.libs
+  NO_DEFAULT_PATH)
+message("Found PROTOBUF_LIBRARIES:" ${PROTOBUF_LIBRARIES})
+find_program(PROTOBUF_PROTOC_EXECUTABLE protoc
+  REQUIRED
+  PATHS ${PROJECT_SOURCE_DIR}/third-party/googlelibraries/protobuf-3.5.2
+  PATH_SUFFIXES src/
+  NO_DEFAULT_PATH)
+message("Found PROTOBUF_PROTOC_EXECUTABLE:" ${PROTOBUF_PROTOC_EXECUTABLE})
```

Comment:
This prints `Found PROTOBUF_PROTOC_EXECUTABLE:/usr/bin/protoc` for me, which later results in error messages about version mismatches.

Reply:
cmake has a tendency to check system paths before those provided in `PATHS`; use `HINTS` instead

Reply:
I use `NO_DEFAULT_PATH` to explicitly disable all other paths